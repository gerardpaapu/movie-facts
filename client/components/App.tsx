import { Routes, Route, Link } from 'react-router-dom'
import MovieList from './MovieList'
import Movie from './Movie'
import Search from './Search'
import Home from './Home'
import Category from './Category'
import CategoryList from './CategoryList'

function App() {
  return (
    <>
      <header className="header">
        <h1>Movie Facts</h1>
      </header>
      <nav>
        <ul>
          <li>
            <Link to="/movie">All movies</Link>
          </li>

          <li>
            <Link to="/category">All categories</Link>
          </li>
          <li>
            <Link to="/search">Search</Link>
          </li>
        </ul>
      </nav>
      <section className="main">
        <Routes>
          <Route index element={<Home />} />
          <Route path="/movie">
            <Route index element={<MovieList />} />
            <Route path=":id" element={<Movie />} />
          </Route>
          <Route path="/category">
            <Route index element={<CategoryList />} />
            <Route path=":id" element={<Category />} />
          </Route>
          <Route path="/search" element={<Search />} />
        </Routes>
      </section>
    </>
  )
}

export default App
