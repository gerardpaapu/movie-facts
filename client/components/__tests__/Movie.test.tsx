/** @jest-environment jsdom */
import nock from 'nock'
import App from '../App'
import { Provider } from 'react-redux'
import { MemoryRouter as Router } from 'react-router-dom'
import store from '../../store'

import { render, waitFor, screen } from '@testing-library/react'

import '@testing-library/jest-dom'

describe('<Movie />', () => {
  it('Loads and shows information about the movie', async () => {
    const scope = nock('http://localhost')
      .get('/api/v1/movies/12?withCategories=true')
      .reply(200, {
        categories: [
          {
            id: 3,
            name: 'Drama',
          },
        ],
        id: 12,
        release_year: 2013,
        title: '12 Years a Slave',
      })

    nock('http://localhost')
      .get('/api/v1/categories')
      .reply(200, [
        {
          id: 3,
          name: 'Drama',
        },
      ])

    const { container } = render(
      <Router initialEntries={['/movie/12']}>
        <Provider store={store}>
          <App />
        </Provider>
      </Router>
    )

    await waitFor(() => expect(scope.isDone()).toBe(true))

    expect(container).toMatchSnapshot()
  })

  it("What happens if the id doesn't exist tho...", async () => {
    const scope = nock('http://localhost')
      .get('/api/v1/movies/23934?withCategories=true')
      .reply(404, undefined) // oops this should be a 404

    nock('http://localhost')
      .get('/api/v1/categories')
      .reply(200, [
        {
          id: 3,
          name: 'Drama',
        },
      ])

    const { container } = render(
      <Router initialEntries={['/movie/23934']}>
        <Provider store={store}>
          <App />
        </Provider>
      </Router>
    )

    //await waitFor(() => expect(scope.isDone()).toBe(true))

    const loadingIndicator = await screen.findByText('Loading...')
    expect(loadingIndicator).toBeVisible()
    expect(container).toMatchInlineSnapshot(`
      <div>
        <header
          class="header"
        >
          <h1>
            Movie Facts
          </h1>
        </header>
        <nav>
          <ul>
            <li>
              <a
                href="/movie"
              >
                All movies
              </a>
            </li>
            <li>
              <a
                href="/category"
              >
                All categories
              </a>
            </li>
            <li>
              <a
                href="/search"
              >
                Search
              </a>
            </li>
          </ul>
        </nav>
        <section
          class="main"
        >
          <p>
            Loading...
          </p>
        </section>
      </div>
    `)
  })
})
