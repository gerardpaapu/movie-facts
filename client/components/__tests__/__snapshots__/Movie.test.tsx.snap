// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Movie /> Loads and shows information about the movie 1`] = `
<div>
  <header
    class="header"
  >
    <h1>
      Movie Facts
    </h1>
  </header>
  <nav>
    <ul>
      <li>
        <a
          href="/movie"
        >
          All movies
        </a>
      </li>
      <li>
        <a
          href="/category"
        >
          All categories
        </a>
      </li>
      <li>
        <a
          href="/search"
        >
          Search
        </a>
      </li>
    </ul>
  </nav>
  <section
    class="main"
  >
    <div>
      <h2>
        12 Years a Slave
         (
        2013
        )
      </h2>
      <form
        aria-label="Add category to movie"
      >
        <label>
          <span>
            New category
          </span>
          <select>
            <option>
              Choose a category
            </option>
          </select>
        </label>
        <button
          type="submit"
        >
          Submit
        </button>
      </form>
      <ul>
        <li>
          <button
            aria-label="delete category Drama"
            value="3"
          >
            delete
          </button>
          <a
            href="/category/3"
          >
            Drama
          </a>
        </li>
      </ul>
    </div>
  </section>
</div>
`;

exports[`<Movie /> What happens if the id doesn't exist tho... 1`] = `
<div>
  <header
    class="header"
  >
    <h1>
      Movie Facts
    </h1>
  </header>
  <nav>
    <ul>
      <li>
        <a
          href="/movie"
        >
          All movies
        </a>
      </li>
      <li>
        <a
          href="/category"
        >
          All categories
        </a>
      </li>
      <li>
        <a
          href="/search"
        >
          Search
        </a>
      </li>
    </ul>
  </nav>
  <section
    class="main"
  >
    <p>
      Failed: 
      Error: Not Found
    </p>
  </section>
</div>
`;
